default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
+
+    growth = 2 # maximum population growth has higher priority.
+    production_boost = 1.5 # production boosts have higher priority. This include buildings that boosts production
+    research_boost = 2 # research boosts have higher priority. This include buildings that boosts research
+    meter_boost = 1
+    shield = 1.25 # shields are more powerful than armors or weapons
+    defensive = 2 if foAI.foAIstate.aggression <= fo.aggression.cautious else (1 if fo.currentTurn() > 10 else 0.5)
+    supply = 2 if ColonisationAI.galaxy_is_sparse() else 0.5
+
+    red_star = len(AIstate.empireStars.get(fo.starType.red, [])) != 0
+    bright_star = len(AIstate.empireStars.get(fo.starType.blue, [])) != 0
+    blackhole = len(AIstate.empireStars.get(fo.starType.blackHole, [])) != 0
+    neutron = len(AIstate.empireStars.get(fo.starType.neutron , [])) != 0
+
+    hull = 1
+    offtrack_hull = 0.05
+    offtrack_subhull = 0.25
+    # select one hull and specialize it, but some AI may want to have more hulls, by random
+    org = hull if index & 1 == 0 or (index + 97) % 227 < 21 else offtrack_hull
+    robotic = hull if index & 1 == 1 or (index + 107) % 229 < 22 else offtrack_hull
+    if ColonisationAI.got_ast:
+        extra = (index + 113) % 223 < 11
+        asteroid = hull if index & 2 == 0 or extra else offtrack_hull
+        if asteroid == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+    else:
+        asteroid = offtrack_hull
+    if bright_star or blackhole:
+        extra = (index + 103) % 211 < 10
+        energy = hull if index & 4 == 0 or extra else offtrack_hull
+        if energy == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+            asteroid = offtrack_hull
+    else:
+        energy = offtrack_hull
+
+    # further specialization for robotic hulls
+    contgrav = hull if robotic == hull and index & 8 == 0 else offtrack_subhull
+    nanorobo = hull if robotic == hull and index & 8 == 1 else offtrack_subhull
+    flux = hull if robotic == hull and (index + 17) % 43 < 22 else offtrack_subhull
+    # further specialization for asteroid hulls
+    astheavy = hull if asteroid == hull and index % 3 == 0 else offtrack_subhull
+    astswarm = hull if asteroid == hull and index % 3 == 1 else offtrack_subhull
+    astcamo = hull if asteroid == hull and index % 3 == 2 else offtrack_subhull
+    # further specialization for organic hulls
+    orgneural = hull if org == hull and index & 8 == 0 else offtrack_subhull
+    orgraven = hull if org == hull and index & 8 == 1 else offtrack_subhull
+    orgendosym = hull if org == hull and (index + 17) % 43 < 22 else offtrack_subhull
+    # further specialization for energy hulls
+    energyfrac = hull if energy == hull and index & 8 == 0 else offtrack_subhull
+    energymag = hull if energy == hull and index & 8 == 1 else offtrack_subhull
```

Comment:
We probably do not want to use a random decision making for what hulls we research but actually base them on the strength of the hulls. We have some module (ShipDesignAI) which can reasonably well estimate the strength of designs. It was intended to extend that module to be able to estimate the use of unlocking certain parts/hulls, i.e. by how much they change the "rating" of our best designs. 

I see no reason to go for a subpar hull in general so randomizing the process seems not to be too useful (though a small random bias could be applied to choose between hulls of comparable strength I suppose). For example, the camo hulls are pretty much unusable right now (at least for the AI). Going for them seems not very useful and the Asteroid swarms are also a really subpar hull to use without some good flagships.